sdk-lib/src/main/java/cash/z/ecc/android/sdk/SdkSynchronizer.kt

```diff
@@ -194,15 +202,49 @@ class SdkSynchronizer private constructor(
         }
     }
 
-    private val _status = MutableStateFlow<Synchronizer.Status>(DISCONNECTED)
-    private val _exchangeRateUsd = MutableStateFlow<Pair<BigDecimal, Instant>?>(null)
+    private val _status = MutableStateFlow(DISCONNECTED)
 
-    var coroutineScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
+    val coroutineScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
 
     override val orchardBalances = processor.orchardBalances.asStateFlow()
     override val saplingBalances = processor.saplingBalances.asStateFlow()
     override val transparentBalance = processor.transparentBalance.asStateFlow()
-    override val exchangeRateUsd = _exchangeRateUsd.asStateFlow()
+
+    private val refreshExchangeRateUsd = MutableSharedFlow<Unit>(replay = 1).apply { tryEmit(Unit) }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override val exchangeRateUsd =
+        channelFlow {
+            var lastValue = ObserveFiatCurrencyResult()
+            refreshExchangeRateUsd
+                .flatMapLatest {
+                    flow {
+                        emit(lastValue.copy(isLoading = true))
+                        lastValue =
+                            when (val result = fetchExchangeChangeUsd()) {
+                                is FetchFiatCurrencyResult.Error -> lastValue.copy(isLoading = false)
+
+                                is FetchFiatCurrencyResult.Success ->
+                                    lastValue.copy(
+                                        isLoading = false,
+                                        currencyConversion = result.currencyConversion
+                                    )
```

Comment:
And then this is where we update older valid data (or initialize the data after the first successful refresh)?

Reply:
yes exactly :) 

We always show latest data and let the client decide whether it's stale or not and how to handle it. That means when we encounter an error during fetch we do not invalidate the previous data.